packages/appium/lib/extension/manifest.js

```diff
@@ -0,0 +1,564 @@
+   * @param {string} pkgPath
+   * @returns {void}
+   */
+  addExtensionFromPackage (pkgJson, pkgPath) {
+    if (!isExtension(pkgJson)) {
+      return;
+    }
+
+    /**
+     * @type {InternalData}
+     */
+    const internal = {
+      pkgName: pkgJson.name,
+      version: pkgJson.version,
+      installType: INSTALL_TYPE_NPM,
+      installSpec: `${pkgJson.name}@${pkgJson.version}`,
+    };
+
+    if (isDriver(pkgJson)) {
+      if (!this._data.drivers[pkgJson.appium.driverName]) {
+        this._data.drivers[pkgJson.appium.driverName] = {
+          ..._.omit(pkgJson.appium, 'driverName'),
+          ...internal,
+        };
+      }
+    } else if (isPlugin(pkgJson)) {
+      if (!this._data.plugins[pkgJson.appium.pluginName]) {
+        this._data.plugins[pkgJson.appium.pluginName] = {
+          ..._.omit(pkgJson.appium, 'pluginName'),
+          ...internal,
+        };
+      }
+    } else {
+      throw new TypeError(
+        `The extension in ${path.dirname(
+          pkgPath,
+        )} is neither a valid driver nor a valid plugin.`,
+      );
+    }
+  }
+
+  /**
+   * Adds an extension to the manifest as was installed by the `appium` CLI.  It determines the
+   * `extData`, `extType`, `extName` itself.
+   *
+   * See {@link Manifest.addExtensionFromPackage} for adding an extension from an on-disk package.
+   * @template {ExtensionType} ExtType
+   * @param {ExtType} extType
+   * @param {string} extName
+   * @param {ExtData<ExtType>} extData
+   * @returns {void}
+   */
+  addExtension (extType, extName, extData) {
+    this._data[`${extType}s`][extName] = extData;
+  }
+
+  /**
+   * Returns the APPIUM_HOME path
+   */
+  get appiumHome () {
+    return this._appiumHome;
+  }
+
+  /**
+   * Returns the path to the manifest file
+   */
+  get manifestPath () {
+    return this._manifestPath;
+  }
+
+  /**
+   * Returns extension data for a particular type.
+   *
+   * @template {ExtensionType} ExtType
+   * @param {ExtType} extType
+   * @returns {ExtRecord<ExtType>}
+   */
+  getExtensionData (extType) {
+    return this._data[/** @type {string} */ (`${extType}s`)];
+  }
+
+  /**
+   * Reads manifest from disk and _overwrites_ the internal data.
+   *
+   * If the manifest does not exist on disk, an {@link INITIAL_MANIFEST_DATA "empty"} manifest file will be created.
+   *
+   * If `APPIUM_HOME` contains a `package.json` with an `appium` dependency, then a hash of the `package.json` will be taken. If this hash differs from the last hash, the contents of `APPIUM_HOME/node_modules` will be scanned for extensions that may have been installed outside of the `appium` CLI.  Any found extensions will be added to the manifest file, and if so, the manifest file will be written to disk.
+   *
+   * Only one read operation should happen at a time.  This is controlled via the {@link Manifest._reading} property.
+   * @returns {Promise<ManifestData>} The data
+   */
+  async read () {
+    if (this._reading) {
```

Comment:
I prefer to use bluebird promises. they have convenience methods like http://bluebirdjs.com/docs/api/ispending.html, which helps to avoid nullifying the value

Reply:
I don't really need anything bluebird provides here and am not using the global `Promise`...  Anyway, if we want to enforce an explicit cast of all promises to bluebird, we should put that in our eslint config.

Reply:
I usually don't really care what type of promise it is. If we need special bluebird features we could always safely call B.resolve(...) on the initial promise and then safely use its features.